x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -0,0 +1,260 @@
+import org.opensaml.security.x509.X509Credential;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.DocumentBuilder;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.security.Signature;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.Inflater;
+import java.util.zip.InflaterInputStream;
+
+import static org.opensaml.saml.saml2.core.NameIDType.UNSPECIFIED;
+
+
+/*
+ * Processes a SAML AuthnRequest, validates it and extracts necessary information
+ */
+public class SamlAuthnRequestValidator {
+
+    private final SamlIdentityProvider idp;
+    private final Logger logger = LogManager.getLogger(SamlAuthnRequestValidator.class);
+    private static final String[] XSD_FILES = new String[]{"/org/elasticsearch/xpack/idp/saml/support/saml-schema-protocol-2.0.xsd",
+        "/org/elasticsearch/xpack/idp/saml/support/saml-schema-assertion-2.0.xsd",
+        "/org/elasticsearch/xpack/idp/saml/support/xenc-schema.xsd",
+        "/org/elasticsearch/xpack/idp/saml/support/xmldsig-core-schema.xsd"};
+
+    private static final ThreadLocal<DocumentBuilder> THREAD_LOCAL_DOCUMENT_BUILDER = ThreadLocal.withInitial(() -> {
+        try {
+            return SamlUtils.getHardenedBuilder(XSD_FILES);
+        } catch (Exception e) {
+            throw new ElasticsearchSecurityException("Could not load XSD schema file", e);
+        }
+    });
+
+    public SamlAuthnRequestValidator(SamlIdentityProvider idp) {
+        SamlUtils.initialize();
+        this.idp = idp;
+    }
+
+    public SamlValidateAuthnRequestResponse processQueryString(String queryString) {
+        final List<String> parameterList = Arrays.asList(queryString.split("&"));
+        final Map<String, String> parameters = new HashMap<>();
+        if (parameterList.isEmpty()) {
+            throw new IllegalArgumentException("Invalid Authentication Request query string");
+        }
+        RestUtils.decodeQueryString(queryString, 0, parameters);
+        logger.trace(new ParameterizedMessage("Parsed the following parameters from the query string: {}", parameters));
+        // We either expect at least a single parameter named SAMLRequest or at least 3 ( SAMLRequest, SigAlg, Signature )
+        final String samlRequest = parameters.get("SAMLRequest");
+        final String relayState = parameters.get("RelayState");
+        final String sigAlg = parameters.get("SigAlg");
+        final String signature = parameters.get("Signature");
+        if (null == samlRequest) {
+            logAndThrow(() -> new ParameterizedMessage("Query string [{}] does not contain a SAMLRequest parameter", queryString));
```

Comment:
I don't understand why we have a lambda here. Can't we just pass a `ParameterizedMessage` to logAndThrow? We know the message will be instantiated since it's going to be thrown.

Reply:
True, I carried this over from previous version where I wouldn't log_and_throw and it made sense then